packages/datadog-plugin-graphql/src/index.js

```diff
@@ -392,11 +396,39 @@ function getOperation (document, operationName) {
   }
 }
 
+function getFragments (document) {
+  if (!document || !Array.isArray(document.definitions)) {
+    return []
+  }
+
+  return document.definitions.reduce((acc, d) => {
+    if (d.kind !== 'FragmentDefinition') return acc
+    return {
+      ...acc,
+      [d.name.value]: d
+    }
+  }, {})
+}
+
+function getExecutionContext (args, res) {
```

Comment:
This should not be necessary since `graphql` already does this. Using what is provided by `graphql` will also avoid duplicating this logic.

Reply:
The plugin does not currently depend on graphql. So, I tried to move the dependency from dev to regular. When I did that and imported `buildExecutionContext` from `graphql/execution/execute`, it broke the tests:

```
Error: Cannot use GraphQLScalarType "String" from another module or realm.

Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.

https://yarnpkg.com/en/docs/selective-version-resolutions

Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.
```

The tests are run against 2 different versions of graphql. I'm not familiar with these kinds of setups and I suspect that could be the root cause of the failure. Can you help me resolve the issue?

Reply:
Since it's exported from the module, it should be possible to grab it from the plugin directly. It should be available at `execute.buildExecutionContext`. In that sense it's not required to require it externally since it's already passed to the plugin.

However, doing so means you would end up with a new execution context that is different than the one provided to the rest of the library internally. It would also mean that the execution context is computed twice, which has an overhead.

Unfortunately, it doesn't look like the internal execution context is available anywhere outside of the `execution/execute.js` file, meaning there is no way to intercept it. Since the execution args may be missing some fields, I think at this point the only possible option will be to pass a custom object that includes any missing field that were added by the tracer. In my opinion the important ones would be everything from the args except the field resolver and type resolver.

In short, because it is unfortunately impossible to use the execution context, we will have to use the args, but augment it when fields are missing to ensure there are references to these fields in the plugin. Back to args it is 😄

Reply:
Actually it looks like the missing pieces are still available eventually on the `info` object passed to resolvers. So in theory is would be possible to combine `args` and `info` to reconstruct the context. I wonder if this should be done at the execute hook level, or if it should be up to the user to combine 2 hooks (execute + resolve) to get the information.

Reply:
> ...we will have to use the args, but augment it when fields are missing...

Do I understand correctly that what we want now is an ExecutionArgs-like object with most of its fields being required?

```ts
// Original ExecutionArgs type
export type ExecutionArgs = {
  schema: GraphQLSchema,
  document: DocumentNode,
  rootValue?: mixed,
  contextValue?: mixed,
  variableValues?: ?{ +[variable: string]: mixed, ... },
  operationName?: ?string,
  fieldResolver?: ?GraphQLFieldResolver<any, any>,
  typeResolver?: ?GraphQLTypeResolver<any, any>,
};

// What we want:
export type ExecutionArgsButBetter = {
  schema: GraphQLSchema,
  document: DocumentNode,
  rootValue: mixed,
  contextValue: mixed,
  variableValues: { [variable: string]: mixed, ... },
  operationName: string,
  fieldResolver?: ?GraphQLFieldResolver<any, any>,
  typeResolver?: ?GraphQLTypeResolver<any, any>,
};
```
> So in theory is would be possible to combine args and info to reconstruct the context. I wonder if this should be done at the execute hook level

How could the context be reconstructed in the execute hook? The `info` object is only available to the wrapped field resolver. It gets passed into the resolver somewhere outside of the scope of the execution wrapper.

> or if it should be up to the user to combine 2 hooks (execute + resolve) to get the information.

I'm not sure I understand how that could be achieved. Could you give an example?

Reply:
> Do I understand correctly that what we want now is an ExecutionArgs-like object with most of its fields being required?

Correct. However, for now I would probably only include common fields that are mutable within the operation to be safe. The type definition would then look something like this:

```ts
export type ExecutionContext = {
  rootValue: any,
  contextValue: any,
  variableValues: any
};
```

Most of the other fields are more difficult to interact with or have limited relevance. I prefer to keep it limited for now since it's still unclear if the args is really the only option.

> How could the context be reconstructed in the execute hook? The info object is only available to the wrapped field resolver. It gets passed into the resolver somewhere outside of the scope of the execution wrapper.

That could be a problem, but not necessarily depending on how it's handled internally. The root resolver should have access to the execution context, but I don't know if there are any guarantees on the presence of at least one resolver. That's something to look into, but since it's more involved I think it's best to try to do it outside the scope of this PR if possible.

> I'm not sure I understand how that could be achieved. Could you give an example?

The following `ExecutionContext` fields are available on the `ExecutionArgs` provided by the user:

* schema
* rootValue (may be undefined)
* contextValue (may be undefined)
* variableValues (may be undefined)
* fieldResolver (may be undefined but is not needed)
* typeResolver (may be undefined but is not needed)

The following `ExecutionContext` fields are available on the `GraphQLResolveInfo` passed to resolvers:

* schema
* operation
* fragments
* rootValue
* variableValues
* operation

By combining the fields from both, it's possible to create an object that is identical to the internal `ExecutionContext` without any additional processing. However, if this is not handled by the hook implementations, it means a user would theoretically have to combine an `execute` hook and `resolve` hook to get all of this information in a single place.

Sorry if this whole thing is getting a bit out of hand. I expected the hooks to be easier to implement with the assumption that the context was exposed by `graphql`.

Reply:
Thank you, @rochdev. I still don't understand the part about combining the fields from 2 different hooks. Nevertheless, I tried to address the requested change in https://github.com/DataDog/dd-trace-js/pull/909/commits/eab4f7f352f7959c8778b356a08af981df1b3efc. Let me know if that matches what you described in one of the comments above 🙂

Reply:
Thank you @vecerek for persevering on this PR, even though it turned out to be harder than expected ❤️

Reply:
@vecerek It definitely matches what I had in mind. However I now realize that `rootValue` can be undefined and `variableValues` is different between `ExecutionArgs` and `ExecutionContext`. At this point I have to capitulate and agree that `ExecutionArgs` is the only way forward. I still think `ExecutionContext` would have been the superior option but it's way too complicated to support a similar object. For the type definition, I would say to do similar to the current `ExecutionContext` type but renamed and with all the keys that are on the actual `ExecutionArgs`, with the same required/optional flag but with everything as type `any`.

Sorry for the runaround. This should be a small change and then it can be merged.

Reply:
Addressed in https://github.com/DataDog/dd-trace-js/pull/909/commits/f8dbcf9ed23f359ffaa6e15f47214702aa9777a2. I hope I didn't forget about anything, unit tests are passing, shouldn't be too bad 😄

Reply:
Should I squash the commits, or would you prefer to leave them for historical reasons? 🙂

Reply:
I'll squash the commits when merging to master, so it's your preference if you want to keep the history in the PR.

Reply:
Let's keep it then.